### PR TITLE
chore: add response logging support for router delivery destinations

### PR DESCRIPTION
**Pull request by sanpj2292**
# Description

Currently, some destinations send events directly from the router, while others send events through transformer. When customers request specific information for debugging, the data is often not readily available. This PR aims to address that issue.

We propose adding logging capabilities for these destinations. The logs will capture the destination's response body, status code, and headers.

This feature will be configurable, requiring the service owner to specify the destType, destinationId, or workspaceId. When an event's data matches any of these criteria, the response information will be logged.

**Note**: The service owner must enable this logging feature for it to take effect.

## Linear Ticket

Resolves INT-2217

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: router/network.go [hunk 1]
```diff
@@ -243,6 +253,36 @@ func (network *netHandle) SendPost(ctx context.Context, structData integrations.
...
+
+	headerBytes, err := json.Marshal(resp.headers)
+	if err != nil {
+		n.logger.Debugn(fmt.Sprintf("header marshal failure: %s", err.Error()),
+			obskit.DestinationID(destInfo.ID),
+			obskit.DestinationType(destInfo.DefinitionName),
+			obskit.WorkspaceID(destInfo.WorkspaceID),
+		)
+	}
+	if destID == destInfo.ID || workspaceID == destInfo.WorkspaceID || destType == destInfo.DefinitionName {
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply router/network.go:274]: Log for all events as we would like to know the request IDs to share the information with users when needed

[the remark this answers is not in the document]
Comment (koladilip) [reply router/network.go:274]: won't this effect rudder-server performance?

[the remark this answers is not in the document]
Comment (sanpj2292) [reply router/network.go:274]: Logging the requests would definitely affect functionality of any service but the need of the hour is that we need to supply the request information details to users when needed. Let me know if there is any other way of tackling this

[the remark this answers is not in the document]
Comment (koladilip) [reply router/network.go:274]: @rudderlabs/server_team is it ok if we enable logging for some duration and log all event responses from the destination

[the remark this answers is not in the document]
Comment (Sidddddarth) [reply router/network.go:274]: > any other way of tackling this

Admittedly without the entire context, it does look like a good candidate for reporting data.
If we're okay with sampled data, going thru reporting data for `sample response` with all the mentioned filters(`destType, destinationId, or workspaceId`) should be good imo.

File: router/network.go [hunk 2]
```diff
@@ -243,6 +253,36 @@ func (network *netHandle) SendPost(ctx context.Context, structData integrations.
...
+	destType := config.GetReloadableStringVar("", "Router.Network.ResponseLogger.destType").Load()
+	workspaceID := config.GetReloadableStringVar("", "Router.Network.ResponseLogger.workspaceID").Load()
+
+	headerBytes, err := json.Marshal(resp.headers)
+	if err != nil {
+		n.logger.Debugn(fmt.Sprintf("header marshal failure: %s", err.Error()),
+			obskit.DestinationID(destInfo.ID),
+			obskit.DestinationType(destInfo.DefinitionName),
+			obskit.WorkspaceID(destInfo.WorkspaceID),
+		)
```
Comment (lvrach) [inline router/network.go:272]:
```suggestion
		n.logger.Debugn("header marshal failure",
		    obskit.Error(err),
			obskit.DestinationID(destInfo.ID),
			obskit.DestinationType(destInfo.DefinitionName),
			obskit.WorkspaceID(destInfo.WorkspaceID),
		)
```

also, is the Debug the appropriate level here? I would go with a Warn

File: router/network.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sanpj2292) [reply router/network.go:274]: Here, we are interested in getting information about the destination's response(when delivered via router) like requestIds, the actual response that we got from destination, any other information that destination sent us.
I don't think sampling actually serves the requirement.

Any user might need these information for a specific set of time range & they would need all the information & not samples.

File: router/network.go (hunk 2 above)
Comment (sanpj2292) [reply router/network.go]: Makes sense. This is done.

File: router/network.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ktgowtham) [reply router/network.go:275]: @sanpj2292 , I don't think it's a great idea to log each and every response. Also, polluting code and configurations just for logging purposes. 
Also, I do see we're logging res body which probably might contain PII information which we should avoid logging.
Can you share any customer requests where customers needed this info?
Is it possible to explore the path Siddharth mentioned above? may be we can enrich reporting sampled data with some more info?

**Pull request closed without merging**